NURBS patch quadrature

dylan-copeland opened the pull request:
This PR introduces new features for numerical integration optimized on NURBS patches rather than elements, for NURBS spaces. Integration rules can be defined in various ways on patches. One option is to compute reduced (sparse), approximate integration rules on patches, using a new NNLS (non-negative least squares) solver for the under-determined systems consisting of the constraints defined by each 1D rule in a tensor product rule applied to the 1D basis functions. For patches, there can be many more basis functions than integration points, giving the potential for reduction in the number of points.

Sample runs show that considerable speed-up is obtained from using patch-wise integration rather than element-wise, and reduced rules give further speed-up.

There is an assumption that the NURBS weights are all 1, which is common in NURBS meshes but not always a valid assumption.

Tasks for future work include
- parallel support,
- GPU support for NURBS PA kernels,
- diagonal assembly (for Jacobi preconditioning),
- support for NURBS with non-unit weights,
- more robust reduced rule computation.

The issue of ensuring "more robust reduced rule computation" is that the computation of reduced rules with NNLS fails in some cases. In order for the system to even be well-posed, the FEM order (affecting the number of basis functions) and integration order (affecting the number of integration points) must be valid choices. When invalid choices are made, the code will fail, and it would be better to simply use the full integration rule.

In `NURBSExtension`, the new data members `p_dof` and `patch_ijk` are set by `NURBSExtension::Generate3DPatchDofTable()` but are never used in this PR. They seem worth keeping as potentially useful data.

A couple of functions are included, which are inefficient and not used. They were written initially as simple implementations, later optimized in the functions that replaced them. The original, inefficient versions can be removed before this PR is merged. The only reason to include them is to help readers understand how the patch-wise assembly works. Those functions are `AssemblePatchMatrix_simpleButInefficient` and `AddMultPatchPA_inefficient`.<!--GHEX{"author":"dylan-copeland","assignment":"2023-01-08T22:56:12","editor":"tzanio","id":3088,"reviewers":["mlstowell","v-dobrev","psocratis"],"merge":"2023-08-22T22:53:04","approval":"2023-08-13T21:49:46"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3088](https://github.com/mfem/mfem/pull/3088) | @dylan-copeland | @tzanio | @mlstowell + @v-dobrev + @psocratis | 1/8/23 | 8/13/23 | 8/22/23 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

IdoAkkerman commented:
I can confirm that there are large gains te be made. So this PR seems very good news.
It seems quite involved so I will try to unpack it carefully.

mlstowell commented on linalg/solvers.hpp:
```diff
@@ -1199,6 +1199,98 @@ class AuxSpaceSmoother : public Solver
+
+   bool NNLS_qrres_on_;
+   QRresidualMode qr_residual_mode_;
+   bool normalize_const_;
+   bool QR_reduce_const_;
```
Two unused variables

mlstowell commented on miniapps/nurbs/nurbs_patch_ex1.cpp:
```diff
@@ -0,0 +1,340 @@
+//
+// Compile with: make nurbs_patch_ex1
+//
+// Sample runs:  nurbs_patch_ex1 -incdeg 3 -ref 2 -iro 8 -patcha
+//               nurbs_patch_ex1 -incdeg 3 -ref 2 -iro 8 -patcha -pa
+//               nurbs_patch_ex1 -incdeg 3 -ref 2 -iro 8 -patcha -fint
```
All three sample runs crash my GLVis with the same error:
```NURBSFECollection: unknown geometry type.```

I just noticed `nurbs_ex1` also produces this error so this probably is not caused by this PR.

mlstowell replied:
It looks like this error was caused by something I changed in PR #3349 a couple months ago. I'll put together a small bugfix PR to correct it.

mlstowell replied:
See PR #3447 for a potential fix.

mlstowell commented on fem/bilininteg.cpp:
```diff
@@ -832,15 +852,31 @@ void DiffusionIntegrator::AssembleElementMatrix
+   const NURBSFiniteElement *NURBSFE =
+      dynamic_cast<const NURBSFiniteElement *>(&el);
+
+   IntegrationRule *irn = nullptr;
```
Why not just use `ir` and avoid the two checks for `irn` below?

dylan-copeland replied:
It would be better to use only one `IntegrationRule*`, but `ir` has to be `const`. I thought `ir` could not be changed. Since you mentioned this, I realize it can be changed (because the `const` is before the type). This C++ feature can be confusing.

mlstowell commented on fem/bilininteg.hpp:
```diff
@@ -61,6 +63,8 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
    virtual void AssemblePA(const FiniteElementSpace &trial_fes,
                            const FiniteElementSpace &test_fes);
 
+   virtual void AssembleNURBSPA(const FiniteElementSpace &fes);
```
Should add a brief comment

mlstowell commented on fem/bilininteg.hpp:
```diff
@@ -80,6 +84,8 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
        called. */
    virtual void AddMultPA(const Vector &x, Vector &y) const;
 
+   virtual void AddMultNURBSPA(const Vector&x, Vector&y) const;
```
Another brief comment at least

mlstowell commented on fem/bilininteg.hpp:
```diff
@@ -2094,15 +2106,31 @@ class DiffusionIntegrator: public BilinearFormIntegrator
 
+   // Data for NURBS patch PA
+   int numPatches = 0;
+   std::vector<std::vector<std::vector<std::vector<Vector>>>> reducedWeights;
+   std::vector<std::vector<std::vector<std::vector<std::vector<int>>>>> reducedIDs;
+   std::vector<Array<int>> pQ1D, pD1D;
+   std::vector<std::vector<Array2D<double>>> pB, pG;
+   std::vector<std::vector<std::vector<int>>> pminD, pmaxD, pminQ, pmaxQ, pminDD,
+       pmaxDD;
```
It might be good to define typedefs (within the class) for these nested containers. It would be more readable. Also, it would provide an opportunity to add explanations of the indices being used e.g. "this nested container takes four indices which are ...".

dylan-copeland replied:
I defined a typedef to simplify.

mlstowell commented on fem/bilininteg.hpp:
```diff
@@ -2129,6 +2157,23 @@ class DiffusionIntegrator: public BilinearFormIntegrator
                                        ElementTransformation &Trans,
                                        DenseMatrix &elmat);
 
+   virtual void AssemblePatchMatrix(const int patch, Mesh *mesh,
+                                    SparseMatrix*& smat);
+
+   void AssemblePatchMatrix_fullQuadrature(const int patch, Mesh *mesh,
+                                           SparseMatrix*& smat);
+
+   void AssemblePatchMatrix_reducedQuadrature(const int patch, Mesh *mesh,
+                                              SparseMatrix*& smat);
+
+   void AssemblePatchMatrix_simpleButInefficient(const int patch,
+                                                 Mesh *mesh,
+                                                 DenseMatrix &pmat);
```
Brief comments would be useful. Particularly since "patch" might be used on non-NURBS contexts and therefore be confusing to some readers.

mlstowell commented on fem/fe/fe_nurbs.hpp:
```diff
@@ -56,6 +56,11 @@ class NURBSFiniteElement : public ScalarFiniteElement
    /// Update the NURBSFiniteElement according to the currently set knot vectors
    virtual void         SetOrder   ()         const { }
+
+   void GetIJK(int *IJK) const
```
This could easily lead to segmentation faults but it is consistent with the API of this class. At the very least I think we need to document this method and make the expected length of `IJK` clear.

mlstowell commented on fem/fespace.hpp:
```diff
@@ -667,8 +667,10 @@ class FiniteElementSpace
    virtual DofTransformation *GetBdrElementDofs(int bel,
                                                 Array<int> &dofs) const;
 
+   void GetPatchDofs(int patch, Array<int> &dofs) const;
```
Again I think we need documentation that refers to NURBS patches.

mlstowell commented on fem/intrules.hpp:
```diff
@@ -257,10 +263,112 @@ class IntegrationRule : public Array<IntegrationPoint>
+   void GetIntegrationPointFrom1D(const int patch, int i, int j, int k,
+                                  IntegrationPoint & ip);
+
+   /// Set pointToElem, used by GetPointElement.
+   void SetPointToElement(Mesh const& mesh);
```
This method name and its comment do describe what the method does but without knowing what `pointToElem` contains and how it will be used this name might be too vague. I would suggest something more descriptive like `BuildPatchIJKToElementMap`. The name doesn't need to be that cumbersome but public methods should be somewhat descriptive. I would then include a reference to `GetPointElement` so that users are directed to the documentation provided there.

dylan-copeland replied:
I changed the name to `Finalize`, which is more understandable and generic.

mlstowell commented on mesh/nurbs.hpp:
```diff
@@ -364,6 +364,8 @@ class NURBSExtension
    int GetNTotalDof() const { return NumOfDofs; }
    int GetNDof()      const { return NumOfActiveDofs; }
 
+   void GetPatchKnotVectors(int p, Array<const KnotVector *> &kv) const;
```
Should add a brief comment

mlstowell commented on mesh/nurbs.hpp:
```diff
@@ -403,6 +405,21 @@ class NURBSExtension
    void KnotInsert(Array<KnotVector *> &kv);
    void KnotInsert(Array<Vector *> &kv);
+
+   int GetElementPatch(int elem) const { return el_to_patch[elem]; }
+   void GetElementIJK(int elem, Array<int> & ijk);
+
+   const Array3D<int>& GetPatchDofs(const int patch);
+
+   int NumPatchDofs1D(const int patch, const int dim);
```
Should document these methods

dylan-copeland replied:
I added documentation in every place you pointed out.

JacobLotz commented:
I noticed that I did not finish my review from some time ago. I have hit the finish button just now and I have removed comments which are not relevant anymore.

This looks like a lot of work you have done!

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -2770,6 +2770,120 @@ void NURBSExtension::Generate3DElementDofTable()
+   {
+      Generate3DPatchDofTable();
+   }
+   else
```
Yes, I agree. The 2D case should be relatively simple to implement, but that will have to wait until I have the time and it becomes a priority, or until someone else :) wants to do it.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -2770,6 +2770,120 @@ void NURBSExtension::Generate3DElementDofTable()
+{
+   if (ndof1D.NumRows() <= patch)
+   {
+      GeneratePatchDofTable();
```
This is a possibility, but I avoided it for the following reasons. 

1. There are many constructors that need to call the same function. This can be done if it is clearly advantageous for performance.
2. Only the 3D case is supported so far. We could do the setup only for 3D in the constructors, but `GetPatchDofs` will still have to verify the 3D case. This is not faster than checking whether the table has been set already.
3. The table `patchDofs` has a significant amount of data, and in most use cases of NURBS it is not necessary. So I am taking this lazy construction approach to avoid constructing `patchDofs` when it is not requested.

JacobLotz replied:
Clear to me!

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -403,6 +405,21 @@ class NURBSExtension
+private:
+   std::vector<Array3D<int>> patchDofs;
+   Array2D<int> ndof1D;
+   std::vector<std::vector<std::set<int>>> patch_ijk;
```
Good observation, that `patch_ijk` is never used. It can be removed. I think I initially planned to use it but did not need it, and thought it may be useful in the future. I will document it, and reviewers can give opinions on whether to keep it. As for `Array2D`, the type must be "trivial", e.g. `int`, `double`, etc. There is a class `IntegerSet` in MFEM, but I have not seen it in use and am not sure whether it is as efficient as `std::set`. Maybe someone else can comment on which is preferable.

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -403,6 +405,21 @@ class NURBSExtension
+   int NumPatchDofs1D(const int patch, const int dim);
+
+private:
+   std::vector<Array3D<int>> patchDofs;
```
`mfem::Vector` is only of type `double`, so I think what you mean is to use `mfem::Array`. It can only take trivial types, so `Array<Array3D<int>>` is not possible.

JacobLotz replied:
Clear!

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -2770,6 +2770,120 @@ void NURBSExtension::Generate3DElementDofTable()
+
+const Array3D<int>& NURBSExtension::GetPatchDofs(const int patch)
+{
+   if ((int) patchDofs.size() <= patch)
```
Right, the casting is to make the types the same. The only reason I would do is to fix a compiler warning, but I can't reproduce the compiler warning when the casting is removed. So I just removed it.

dylan-copeland replied:
Now I remember, the CI fails with an error, although I cannot reproduce it with my compiler. So the casting has to be done.

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -364,6 +364,8 @@ class NURBSExtension
    int GetNTotalDof() const { return NumOfDofs; }
    int GetNDof()      const { return NumOfActiveDofs; }
 
+   void GetPatchKnotVectors(int p, Array<const KnotVector *> &kv) const;
```
Done.

IdoAkkerman commented:
High-level comments:
1) I like the idea of switching from integration over elements to integration over a patch. I was thinking about changing the integration rule (and ip count)  every element, but that is horrible. This seems like the way to go.

2) I have not looked at the NNLS solver and quadrature rule itself, but I think that will be more a courtesy thing then something else.

3) I am not convinced yet by the PatchDofTable addition to the NURBS class. I will clone the PR and have a closer look, and wait for some clarifying answers.

IdoAkkerman commented:
I think the NNLS solver class might need some tidying up.

I have some questions/comments regarding the NNLS solver:
1) It does not inherent from Solver, opposite to most other solvers. I can see this might be due to  upper and lower limits for the solve  resulting the a-traditional call:
```
  void Solve(const DenseMatrix& matTrans,
              const Vector& rhs_lb,
              const Vector& rhs_ub, Vector& soln);
```
However....
2) In _bilininteg_diffusion_patch.cpp_ there is a call
` void SolveNNLS(DenseMatrix & Gt, Vector const& w, Vector & sol)`
I think this routine needs to move to the NNLS solver class. Moreover, this call can be made analogous to the other solver `Mult` calls.  I think.

3) The interfacing of parameters seems different then on the other solver classes. It seems preferable to make them inline with the other solvers.
Additionally, the delta in _SolveNNLS_ is currently hardwired.

4) I think it would be beneficial to add a unit test to test the NNLS solver.

5) Perhaps the NNLS solver class can be handled in a separate PR? This might streamline the process and get other (solver oriented) reviewers involved.

IdoAkkerman commented:
Similar to the issues raised with the NNLS solver I have some comments on the intrules.
1) In _bilininteg_diffusion_patch.cpp_  there is a routine `GetReducedRule`. I think this shoudl be moved to  _intrules.cpp_
2) The new reduced integration rule  should have a unit test.

IdoAkkerman commented:
The miniapp _./nurbs_patch_ex1_ does not always work. Perhaps I am doing something stupid....
I see what is the issue. 
The -iro value needs to be large enough wrt the -incdeg value. You need soem minimal integration order to make sure the NNLS system has the correct dimensions.

I think this should be better documented, or better yet the iro should be modified to the minimal working order with a warning message perhaps.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -3305,6 +3419,15 @@ void NURBSExtension::Set3DSolutionVector(Vector &coords, int vdim)
    }
 }
 
+void NURBSExtension::GetElementIJK(int elem, Array<int> & ijk)
```
It is now using `GetRow`, but the copy is still necessary, since the data is protected.

IdoAkkerman replied:
Okay. perfect

dylan-copeland commented:
Thank you @IdoAkkerman for your suggestions. You are right that the `NNLS` class could be improved. It is adapted from another code outside MFEM, without the MFEM `Solver` class in mind. Some issues with making it derived from the `Solver` class are (i) as you mentioned, `SolveNNLS` would need to become the `Mult` function, assuming that the upper and lower bounds should be computed as in that function, which limits the generality of usage of `NNLS`; (ii) the operator must be a `DenseMatrix` in order to access entries; (iii) `NormalizeConstraints` scales the rows of the `DenseMatrix`.

These issues are addressed now in the following ways: (i) `SolveNNLS` is now the `Mult` function, and `Solve` is still a public function, so that it can be called for any upper and lower bound functions a user may want to use; (ii) a type cast verifies the type `DenseMatrix`; (iii) instead of scaling the `DenseMatrix`, `NNLS` now stores the row scaling in a vector.

I also added a small unit test for NNLS, but the miniapp is a better test. Note that regression testing on the miniapp sample runs should also ensure NNLS is not broken.

In the case that insufficient integration points are used in NNLS, there is now an error message informing the user to increase the number of points by increasing the order of integration. I don't think we can do more than printing an error message, i.e. we cannot pass this information back to the driver code through the integrator API.

I added some documentation for `GetReducedRule` but did not move it to intrules.cpp, because it is very specific to `DiffusionIntegrator`.

IdoAkkerman commented:
What happens when patch assembly is requested for a non-NURBS mesh? 
I think the assembly will be skipped and an obscure error is given due to it.

Perhaps it is better to omit the option is it can not be used -- plus a message of ignoring the flag.

dylan-copeland commented:
@IdoAkkerman Good point, that the failure for a non-NURBS mesh was not informative. I added an `MFEM_VERIFY` with an error message in `BilinearForm::Assemble`, since that function applies the integrators to a mesh. The integrators themselves do not know what mesh will be used, so we cannot check when defining the integrator.

IdoAkkerman commented:
@dylan-copeland Please inform me if I have overstepped the netiquette, but I changed some of the interfaces with the 
NURBSExtension class. As mentioned before I did not see the need for some of the additions. In closer look there seemed a large overlap with already provided features. I think it helps for future maintenance and extensibility that the code is a concise and uniform as possible. So i hope you see the value of my "addition".

dylan-copeland commented:
Thanks @IdoAkkerman for your interface changes. I agree that your changes are an improvement.

dylan-copeland commented:
Ping @v-dobrev @psocratis

v-dobrev commented:
@dylan-copeland, I'll try to review the PR this week. Can you please resolve the conflicts with `master`?

dylan-copeland commented:
Ping @v-dobrev @psocratis

tzanio reviewed (approved):
Thanks @dylan-copeland, this looks great!

tzanio commented:
@psocratis and @v-dobrev -- are you interested in reviewing this?

psocratis commented:
> @psocratis and @v-dobrev -- are you interested in reviewing this?

Yes I will take a look at it this week

psocratis commented on fem/bilininteg_diffusion_pa.cpp:
```diff
@@ -1738,4 +1762,413 @@ void DiffusionIntegrator::AddMultTransposePA(const Vector &x, Vector &y) const
+
+   // NOTE: the following is adapted from AssemblePatchMatrix_fullQuadrature
+   std::vector<Array3D<double>> grad(dim);
+   // TODO: Can an optimal order of dimensions be determined, for each patch?
```
is this pending to be done in the PR?

dylan-copeland replied:
This is an optimization that could be made in the future, but I did not intend to resolve it in this PR. The issue is that patches could generally have very different numbers of DOFs and quadrature points in the 3 spatial directions. The code only has one loop order over the spatial directions, which may not be optimal in general. Setting an order at run-time would make the code more complicated.

psocratis commented on miniapps/nurbs/nurbs_patch_ex1.cpp:
```diff
@@ -0,0 +1,340 @@
+   //    instead use an isoparametric/isogeometric space.
+   FiniteElementCollection *fec;
+   bool delete_fec;
+   if (order > 0)
```
since there is no option for setting the order should we remove this case?

dylan-copeland replied:
Removed. The positive order case is not currently supported.

psocratis commented on miniapps/nurbs/nurbs_patch_ex1.cpp:
```diff
@@ -0,0 +1,340 @@
+   // 9. Set up the bilinear form a(.,.) on the finite element space
+   //    corresponding to the Laplacian operator -Delta, by adding the Diffusion
+   //    domain integrator.
+
```
empty line

dylan-copeland replied:
It is really unbelievable that I missed this.

psocratis commented on miniapps/nurbs/nurbs_patch_ex1.cpp:
```diff
@@ -0,0 +1,340 @@
+   // any necessary transformations such as: eliminating boundary conditions,
+   // applying conforming constraints for non-conforming AMR, static
+   // condensation, etc.
+   if (static_cond) { a.EnableStaticCondensation(); }
```
Also, we shouldn't have static condensation option, right?

dylan-copeland replied:
Removed.

psocratis commented on miniapps/nurbs/nurbs_patch_ex1.cpp:
```diff
@@ -0,0 +1,340 @@
+                  "Mesh file to use.");
+   args.AddOption(&static_cond, "-sc", "--static-condensation", "-no-sc",
+                  "--no-static-condensation", "Enable static condensation.");
+   args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
```
`-pa` with `-no-patcha` would segfault. Maybe we should add a check!

dylan-copeland replied:
Added a check.

psocratis reviewed (commented):
I have some small comments here and there, but overall this looks very good. I am not so familiar with nurbs so I did the best I could. Btw, what would it take to support this in parallel?

dylan-copeland commented on fem/bilininteg.hpp:
```diff
@@ -2144,6 +2181,28 @@ class DiffusionIntegrator: public BilinearFormIntegrator
+                                              SparseMatrix*& smat);
+
+   /// TODO: remove this before merging PR?
+   void AssemblePatchMatrix_simpleButInefficient(const int patch,
```
Initially, I thought it may be helpful to keep these inefficient versions, as they are simpler to read and understand, as a stepping stone to understanding the other optimized integrators (which are rather cryptic). I have removed them now, because I suppose they have too much code to keep just for educational purposes.

dylan-copeland commented on fem/bilinearform_ext.cpp:
```diff
@@ -386,13 +393,39 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
+   }
+
+   MFEM_VERIFY(!(somePatchwise && !allPatchwise),
+               "All or none should be patchwise");
```
Done.

dylan-copeland commented on linalg/solvers.cpp:
```diff
@@ -3544,4 +3544,841 @@ void AuxSpaceSmoother::Mult(const Vector &x, Vector &y, bool transpose) const
+   : Solver(0), mat(nullptr), const_tol_(const_tol), min_nnz_(min_nnz),
+     max_nnz_(max_nnz), verbosity_(verbosity),
+     res_change_termination_tol_(res_change_termination_tol),
+     zero_tol_(zero_tol), rhs_delta_(rhs_delta), normalize_(normalize),
```
Reordered.

dylan-copeland commented on linalg/solvers.cpp:
```diff
@@ -3544,4 +3544,841 @@ void AuxSpaceSmoother::Mult(const Vector &x, Vector &y, bool transpose) const
+{
+   // Scale everything so that rescaled half gap is the same for all constraints
+   const int m = mat->NumRows();
+   const int n = mat->NumCols();
```
Removed.

dylan-copeland commented on fem/bilininteg.hpp:
```diff
@@ -2178,6 +2237,13 @@ class DiffusionIntegrator: public BilinearFormIntegrator
 
+   virtual void AddMultNURBSPA(const Vector&, Vector&) const;
+
+   void AddMultPatchPA_inefficient(const int patch, const Vector &x,
```
Removed.

dylan-copeland commented:
> I have some small comments here and there, but overall this looks very good. I am not so familiar with nurbs so I did the best I could. Btw, what would it take to support this in parallel?

Thanks @psocratis for your review. In addition to parallel support, there are several other future tasks listed in the PR description. I am not sure yet what that implementation will entail. It will be simpler if NURBS patches are not partitioned in parallel, in which case it may be straightforward.

psocratis commented:
@dylan-copeland , there is a test failing due to the latest commit

v-dobrev commented on fem/bilinearform.cpp:
```diff
@@ -421,23 +422,30 @@ void BilinearForm::Assemble(int skip_zeros)
+
+         if (domain_integs[k]->Patchwise())
+         {
+            MFEM_VERIFY(mesh->NURBSext, "Patchwise integration requires a "
+                        << "NURBS mesh");
```
I suspect the patch integration also requires a NURBS FE space -- we can check for that too -- check that `fes->GetNURBSext() != nullptr`:
```suggestion
            MFEM_VERIFY(mesh->NURBSext, "Patchwise integration requires a "
                        << "NURBS mesh");
            MFEM_VERIFY(fes->GetNURBSext(), "Patchwise integration requires a "
                        << "NURBS FE space");
```

v-dobrev commented on fem/bilinearform.cpp:
```diff
@@ -479,6 +488,43 @@ void BilinearForm::Assemble(int skip_zeros)
       }
+
+      // Patch-wise integration
+      if (mesh->NURBSext)
```
Similar to above, it's probably better to check for NURBS FE space:
```suggestion
      if (fes->GetNURBSext())
```

v-dobrev commented on fem/bilinearform_ext.cpp:
```diff
@@ -295,7 +295,14 @@ void PABilinearFormExtension::Assemble()
    for (int i = 0; i < integratorCount; ++i)
    {
-      integrators[i]->AssemblePA(*a->FESpace());
+      if (a->FESpace()->GetMesh()->NURBSext && integrators[i]->Patchwise())
```
Check for NURBS FE space, instead of just checking for NURBS mesh:
```suggestion
      if (a->FESpace()->GetNURBSext() && integrators[i]->Patchwise())
```

v-dobrev replied:
It maybe better to dispatch to `AssembleNURBSPA` or `AssemblePA` just based on `integrators[i]->Patchwise()` and trigger an error if patch-wise integration is requested but the space is not NURBS FE space. What do you think?

dylan-copeland replied:
Done.

v-dobrev commented:
It looks like we hit another status letter combination in `branch-history` that is not handled: `RA`, see https://github.com/mfem/mfem/actions/runs/5722539954/job/15505752072?pr=3088.

Maybe we should treat it the same way we treat `MA`? @pazner, what do you think?

pazner commented:
> It looks like we hit another status letter combination in `branch-history` that is not handled: `RA`, see https://github.com/mfem/mfem/actions/runs/5722539954/job/15505752072?pr=3088.
> 
> Maybe we should treat it the same way we treat `MA`? @pazner, what do you think?

I think something like this patch should do it:
```diff
diff --git a/tests/scripts/branch-history b/tests/scripts/branch-history
index f0c038df7..5b9539839 100755
--- a/tests/scripts/branch-history
+++ b/tests/scripts/branch-history
@@ -157,6 +157,6 @@ foreach my $sha (@commits) {
         $blob_size += $sz1 < $sz2 ? $sz1 : $sz2; }
-      elsif ($mode eq "AM") {
+      elsif ($mode =~ m/A[MR]/) {
         my $sz2 = int(`git diff -U0 --binary $src2 $dst | gzip -c | wc -c`);
         $blob_size += $sz2; }
-      elsif ($mode eq "MA") {
+      elsif ($mode =~ m/[MR]A/) {
         my $sz1 = int(`git diff -U0 --binary $src1 $dst | gzip -c | wc -c`);
```

v-dobrev commented:
@tzanio, this PR should be ready for merging in `next`.

tzanio commented:
Merged in `next` for testing...
